core/src/deserialize.rs

```diff
@@ -277,88 +291,99 @@ impl<'de> serde::Deserializer<'de> for RichTerm {
         visitor.visit_newtype_struct(self)
     }
 
-    /// Deserialize `RichTerm::Bool`
+    /// Deserialize a boolean value.
     fn deserialize_bool<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
         V: Visitor<'de>,
     {
-        match unwrap_term(self)? {
-            Term::Bool(v) => visitor.visit_bool(v),
-            other => Err(RustDeserializationError::InvalidType {
+        match self.as_inline() {
+            Some(InlineValue::True) => visitor.visit_bool(true),
+            Some(InlineValue::False) => visitor.visit_bool(false),
+            _ => Err(RustDeserializationError::InvalidType {
                 expected: "Bool".to_string(),
-                occurred: other.type_of().unwrap_or_else(|| "Other".to_string()),
+                occurred: self.type_of().unwrap_or("Other").to_owned(),
             }),
         }
     }
 
-    /// Deserialize `RichTerm::Str` as char
+    /// Deserialize a string as a char.
     fn deserialize_char<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
         V: Visitor<'de>,
     {
         self.deserialize_string(visitor)
     }
 
-    /// Deserialize `RichTerm::Str` as str
+    /// Deserialize a string (borrowed).
     fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
         V: Visitor<'de>,
     {
-        self.deserialize_string(visitor)
+        match self.as_string() {
+            Some(s) => visitor.visit_str(s),
+            _ => Err(RustDeserializationError::InvalidType {
+                expected: "Str".to_string(),
```

Comment:
Is this a leftover from when String was Str (and Number was Num)? There are also a couple more below.

Reply:
Indeed, good catch.